moon/controller.py

```diff
@@ -118,14 +122,38 @@ def send_speed_cmd(self, speed, angular_speed):
             self.virtual_bumper.update_desired_speed(speed, angular_speed)
         self.bus.publish('desired_speed', [round(speed*1000), round(math.degrees(angular_speed)*100)])
 
-    def send_request(self, cmd):
-        """Send ROS Service Request form a single place"""
-        self.publish('request', cmd)
-        while True:
-            dt, channel, data = self.listen()
-            if channel == 'response':
-                return data
-            print(dt, 'ignoring', channel)
+    def on_response(self, timestamp, data):
+        intoken, response = data
+        self.requests[intoken] = {
+            "response": response
+            }
+
+    def poll_responses(self, token, callback):
+        while self.time is None:
+            self.sleep(0.1)
+        request_sent = self.time
+        while self.time - request_sent < timedelta(seconds=5):
+            for i in self.requests:
+                if self.requests[token]["response"] is not None:
```

Comment:
I don't see who changes this value to `not None` (searching for "requests" in this diff yield nothing). But the intention seems to be to change it from a different thread in which case this would result in race condition.

Reply:
Thanks, I am open to suggestions here.  There are calls unrelated to movement of the robot (such as fetching the list of assignments from the task server) which make sense to trigger without waiting for the robot to be fully initiated first. However, to manage time-out of such a call, the self.time variable is used which only gets initiated some time into the robot start-up sequence.  There is this code in the solution template:

```
while self.last_position is None or self.yaw is None:
  self.update()  # define self.time
```

which waits for the system to be initiated and that's what this code waits and relies on as well.
 However, if there is a better way to 

1. implement the timeout (without using self.time) or 
2. wait for the self.time to be initiated or 
3. build the event sequence (e.g., only call these non-robot calls after the robot has been set up)

I am happy to change the code accordingly. Thanks.

Reply:
I went for option 3 and now starting am operations only after the robot init sequence has been completed